bpf/bpf_host.c

```diff
@@ -526,6 +526,23 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 			return DROP_INVALID;
 	}
 
+#if defined(NO_REDIRECT) && !defined(ENABLE_REDIRECT_FAST)
```

Comment:
Isn't this only an issue if BPF masquerading is disabled? Does it also happen when KPR is enabled but not BPF masquerading?

I also think the proper place to fix this would be in https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L1451-L1465.

Reply:
> Isn't this only an issue if BPF masquerading is disabled? Does it also happen when KPR is enabled but not BPF masquerading?

The issue is only there when both BPF masquerading and KPR are disabled (i.e. not an issue with BPF masquerading disabled and KPR enabled). Does it mean the guard above is incorrect?

> I also think the proper place to fix this would be in https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L1451-L1465.

What's the advantage of skipping redirect in `bpf_lxc`? That we would do:
```
bpf_host [netdev] -> bpf_lxc -> stack
```

rather than:
```
bpf_host [netdev] -> stack -> bpf_host [cilum_host] -> bpf_lxc
```
skipping the `bpf_host` program attached to `cilium_host`?

Reply:
> The issue is only there when both BPF masquerading and KPR are disabled (i.e. not an issue with BPF masquerading disabled and KPR enabled). Does it mean the guard above is incorrect?

I think so. At least it does not encode "BPF masquerading and KPR are disabled". For that you would need to check `ENABLE_MASQUERADING` and `ENABLE_NODEPORT` I believe.

> What's the advantage of skipping redirect in `bpf_lxc`?

Not bypassing ingress policy enforcement when endpoint routes are disabled :wink: 

`ipv4_local_delivery` will tail call to the destination pod's BPF program and enforce ingress policy. If endpoint routes are disabled, that's the only place where we enforce ingress policies.

Reply:
> I think so. At least it does not encode "BPF masquerading and KPR are disabled". For that you would need to check `ENABLE_MASQUERADING` and `ENABLE_NODEPORT` I believe.

ack, will look into that

> Not bypassing ingress policy enforcement when endpoint routes are disabled wink
> 
> `ipv4_local_delivery` will tail call to the destination pod's BPF program and enforce ingress policy. If endpoint routes are disabled, that's the only place where we enforce ingress policies.

I'm not sure I'm following: if endpoint routes are disabled it means that endpoint traffic should go through `cilium_host` (which will not skip redirect as `from_host` will be `true` in this context), which in turn should tail call to the destination pod's BPF program and enforce ingress policy. Am I missing something?

Reply:
> if endpoint routes are disabled it means that endpoint traffic should go through `cilium_host` (which will not skip redirect as `from_host` will be `true` in this context), which in turn should tail call to the destination pod's BPF program and enforce ingress policy. Am I missing something?

Ah, right :+1: It should work then...

Reply:
Maybe we should just make the `#if defined(ENCAP_IFINDEX) || defined(NO_REDIRECT)` check at the line 479 in your PR independent from `#ifdef ENABLE_NODEPORT`?

Reply:
I run a few tests to confirm when all these macros are defined:

* `--kube-proxy-replacement disabled`:
```
root@k8s1:~# cilium status | egrep 'Masquerading|KubeProxyReplacement'
KubeProxyReplacement:   Disabled
Masquerading:           IPTables [IPv4: Enabled, IPv6: Enabled]
root@k8s1:~# egrep 'ENABLE_NODEPORT |ENABLE_MASQUERADE |NO_REDIRECT ' /var/run/cilium/state/globals/node_config.h
#define NO_REDIRECT 1
```

* `--kube-proxy-replacement strict`:
```
root@k8s1:~# cilium status | egrep 'Masquerading|KubeProxyReplacement'
KubeProxyReplacement:   Strict   [enp0s8 192.168.33.11 fd01::b (Direct Routing)]
Masquerading:           IPTables [IPv4: Enabled, IPv6: Enabled]
root@k8s1:~# egrep 'ENABLE_NODEPORT |ENABLE_MASQUERADE |NO_REDIRECT ' /var/run/cilium/state/globals/node_config.h
#define ENABLE_NODEPORT 1
#define NO_REDIRECT 1
```

*` --kube-proxy-replacement strict --enable-bpf-masquerade`:
```
root@k8s1:~# cilium status | egrep 'Masquerading|KubeProxyReplacement'
KubeProxyReplacement:   Strict   [enp0s8 192.168.33.11 fd01::b (Direct Routing)]
Masquerading:           BPF   [enp0s8]   10.11.0.0/16 [IPv4: Enabled, IPv6: Enabled]
root@k8s1:~# egrep 'ENABLE_NODEPORT |ENABLE_MASQUERADE |NO_REDIRECT ' /var/run/cilium/state/globals/node_config.h
#define ENABLE_MASQUERADE 1
#define ENABLE_NODEPORT 1
#define NO_REDIRECT 1
```

to my understanding this means we should assert both that `NO_REDIRECT` is defined as well as `ENABLE_NODEPORT` is _not_ defined, otherwise we'll skip redirect even in the second and third case (which is not optimal as we are redirecting packets back to the stack when we could avoid that)

Reply:
So shouldn't this rather be: `bool skip_redirect = !is_defined(ENABLE_NODEPORT)` instead of `bool skip_redirect = false;` ?

Also, what about v6, seems missing?

Reply:
>So shouldn't this patch rather be: bool skip_redirect = !is_defined(ENABLE_NODEPORT) instead of bool skip_redirect = false; ? We already have skip_redirect taking care of this. Why it cannot be reused?

If we do like that, we would skip redirect just if `ENABLE_NODEPORT` is not defined (without checking also that `NO_REDIRECT` is set)

> Also, what about v6, seems missing?

(I think I pushed the changes for v6 right before you commented on the PR)